app/controllers/concerns/signature_verification.rb

```diff
@@ -24,16 +24,27 @@ def signature_verification_failure_code
   end
 
   def signature_key_id
-    raw_signature    = request.headers['Signature']
-    signature_params = {}
+    signature_params['keyId']
+  end
 
-    raw_signature.split(',').each do |part|
-      parsed_parts = part.match(/([a-z]+)="([^"]+)"/i)
-      next if parsed_parts.nil? || parsed_parts.size != 3
-      signature_params[parsed_parts[1]] = parsed_parts[2]
-    end
+  def signature_algorithm
+    signature_params['algorithm'] || 'hs2019'
+  end
 
-    signature_params['keyId']
+  def signature_params
+    @signature_params ||= begin
+      raw_signature    = request.headers['Signature']
+      signature_params = {}
+
+      # TODO: this is really approximative... we should fix it with proper parsing
```

Comment:
what would the proper parsing look like here? In which way is this implementation approximate? Can you give examples of params where the "proper parsing" would create a hash that differs from the one generated by this code?

Reply:
signature params are defined in the spec as `sig-param    = token BWS "=" BWS ( token / quoted-string )`
in particular, `token` allows more than alphabetic characters, the right-hand side isn't necessarily a quoted-string, and a quoted-string could include a comma, I think

Reply:
I rewrote the parser with `Parslet`, it now correctly handles token values (in addition to quoted strings), as well as commas and escaped quotes in quoted strings, and optional spaces wherever they are allowed. It is not 100% compliant with the spec, though, as my implementation may allow non-ascii or non-printable characters in quoted strings, contrary to the spec, but HTTP's grammar definitions are a pain, and I think this is close enough.